### ISSUE TITLE
google_bigquery_reservation_assignment does not support all job_type values in BigQuery

### ISSUE THREAD

**Issue by dangermike** [hashicorp/terraform-provider-google#17485]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

### Description

The [google_bigquery_reservation_assignment[(https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_reservation_assignment) type specifies the following `job_type` values:

* JOB_TYPE_UNSPECIFIED
* PIPELINE
* QUERY

However, [BigQuery specifies](https://cloud.google.com/bigquery/docs/reservations-intro#assignments):
* QUERY
* PIPELINE
* BACKGROUND
* ML_EXTERNAL

This renders Terraform unable to model all reservation assignment types.

### New or Affected Resource(s)

* google_bigquery_reservation_assignment

### Potential Terraform Configuration

```tf
resource "google_bigquery_reservation_assignment" "primary" {
  assignee  = "projects/my-project-name"
  job_type = "ML_EXTERNAL"
  reservation = google_bigquery_reservation.basic.id
}
```

**Comment by dangermike** [comment]
These type appear in the [apiv1 docs](https://cloud.google.com/go/docs/reference/cloud.google.com/go/bigquery/latest/reservation/apiv1/reservationpb#cloud_google_com_go_bigquery_reservation_apiv1_reservationpb_Assignment_JobType), but do not appear in the [DCL](https://github.com/GoogleCloudPlatform/declarative-resource-client-library/blob/main/services/google/bigqueryreservation/assignment.yaml). Am I barking up the wrong tree by bringing this request here?

**Comment by dangermike** [comment]
I have reached out to my Google Cloud TAMs to try to get the DCL updated:

---

It looks like the terraform resource for BigQuery reservations ([google_bigquery_reservation_assignment](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_reservation_assignment)) specifies job_type as only JOB_TYPE_UNSPECIFIED, PIPELINE, or QUERY. However, the Google Cloud [documentation](https://cloud.google.com/bigquery/docs/reservations-intro#assignments) and the API documentation ([go](https://cloud.google.com/go/docs/reference/cloud.google.com/go/bigquery/latest/reservation/apiv1/reservationpb#cloud_google_com_go_bigquery_reservation_apiv1_reservationpb_Assignment_JobType), [REST](https://cloud.google.com/bigquery/docs/reference/reservations/rest/v1/projects.locations.reservations.assignments#JobType)) both also include ML_EXTERNAL and BACKGROUND as available assignment types. I have filed an [issue with Terraform](https://github.com/hashicorp/terraform-provider-google/issues/17485), but I think I may be talking to the wrong people. They base their clients on [DCL](https://github.com/GoogleCloudPlatform/declarative-resource-client-library/), and the [DCL it’s based on](https://github.com/GoogleCloudPlatform/declarative-resource-client-library/blob/718ebd0ab5af1565ae5b755b215ea32ea5779433/services/google/bigqueryreservation/assignment.yaml#L98-L108) is missing those enum values! In fact, when you look in the [Terraform provider code](https://github.com/hashicorp/terraform-provider-google/blob/main/google/services/bigqueryreservation/resource_bigquery_reservation_assignment.go), there is a big disclaimer at the top saying that they won’t take any contributions because the whole thing is generated from the DCL.

In short, all signs point to that DCL needing an update. By not having ML_EXTERNAL, the DCL must be older than [August, 2021](https://github.com/googleapis/googleapis/commit/2c7f233a072ce7acf72a1db871a94a7c3e8ad36e#diff-beef70125ad5a16a5e396eaa383b58e9be3d9893f289af94c133641272db0c55R733), though BACKGROUND was only added in [March, 2023](https://github.com/googleapis/googleapis/commit/e4a7ae691c2b3f5448e1d7ab9831aca3e1e6d0ec#diff-beef70125ad5a16a5e396eaa383b58e9be3d9893f289af94c133641272db0c55R862).

Is there any way to get to the right people and poke them for a DCL update? This will impact anyone with ML_EXTERNAL or BACKGROUND reservation assignments they are trying to manage through Terraform.

**Comment by rileykarson** [comment]
We're the owners of both the provider and the DCL, this is the right place to have filed this. Feature triage happens on a weekly cadence is all!

**Comment by dangermike** [comment]
@rileykarson it's been two months. Should I assume that this being on the "Goals" milestone means that this isn't going to happen in the near term?

**Comment by rileykarson** [comment]
Yeah, we don't guarantee specific closure times for most issues. If you're a larger GCP customer there are some addtl channels for input (although getting a guarantee there can be difficult too, subject to Google-internal critieria), from the issue template:

> Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

**Comment by dangermike** [comment]
Thanks, @rileykarson. I wasn't looking for any kind of guarantee, but it seems like this has been pushed into the backlog. I'll ask my TAMs if they can pull the right levers for me.

**Comment by smenke** [comment]
I can confirm this is working for me with BACKGROUND & BACKGROUND_CHANGE_DATA_CAPTURE. The documentation does not reflect that.
